### PR TITLE
various perf optimization and small fixes

### PR DESCRIPTION
**Pull request by heejaechang**
first, fixed one null exception and a race found during this work.
second, added some cancellation token check to make cancellation faster.
third, send big source text as chunk rather than one big string (most of time, source text are sent as memory map file, so this is mostly for active file user is working on)
last, use persistent service for source text checksum. source text checksum is most expansive checksum to calculate so, reuse the checksum if possible.

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @srivatsn @panopticoncentral can you take a look?

**Comment by TyOverby** [comment]
@dotnet-bot test perf please

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState_Checksum.cs [hunk 1]
```diff
@@ -30,15 +37,90 @@ public Task<DocumentStateChecksums> GetStateChecksumsAsync(CancellationToken can
         {
             using (Logger.LogBlock(FunctionId.DocumentState_ComputeChecksumsAsync, FilePath, cancellationToken))
             {
-                var textTask = GetTextAsync(cancellationToken);
+                var textAndVersionTask = GetTextAndVersoinAsync(cancellationToken);
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState_Checksum.cs]: Typo "Versoin"

File: src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs [hunk 2]
```diff
@@ -190,5 +195,205 @@ public static bool AreOnSameLine(this SourceText text, SyntaxToken token1, Synta
...
+
+            // write whole chunks
+            try
+            {
+                var offset = 0;
+                for (var i = 0; i < numberOfChunks; i++)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var count = Math.Min(buffer.Length, length - offset);
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs]: You can probably simplify this code by looping from i to i < numberOfChunks - 1, and then processing the last chunk outside of the loop.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs]: due to this condition, "if ((i < numberOfChunks - 1) || (count == buffer.Length))", I think having just one loop is better.

**Pull request merged**